File: python/docs/examples/filesystem_example/watch_directory_example.py
#!/usr/bin/env python3
"""
Watch Directory Example

This example demonstrates how to use the watch_directory functionality
to monitor file changes in a directory.
"""

import os
import time
import threading
from agentbay import AgentBay
from agentbay.session_params import CreateSessionParams


def main():
    """Main function demonstrating watch_directory usage."""
    
    # Get API key from environment variable
    api_key = os.getenv("AGENTBAY_API_KEY")
    if not api_key:
        print("❌ Please set the AGENTBAY_API_KEY environment variable")
        return
    
    print("🚀 Watch Directory Example")
    print("=" * 50)
    
    # Initialize AgentBay client
    agentbay = AgentBay(api_key=api_key)
    print("✅ AgentBay client initialized")
    
    # Create session with code_latest ImageId
    session_params = CreateSessionParams(image_id="code_latest")
    session_result = agentbay.create(session_params)
    
    if not session_result.success:
        print(f"❌ Failed to create session: {session_result.error_message}")
        return
    
    session = session_result.session
    print(f"✅ Session created: {session.session_id}")
    
    try:
        # Create a test directory to monitor
        test_dir = "/tmp/watch_example"
        print(f"\n📁 Creating test directory: {test_dir}")
        
        create_result = session.file_system.create_directory(test_dir)
        if not create_result.success:
            print(f"❌ Failed to create directory: {create_result.error_message}")
            return
        
        print("✅ Test directory created")
        
        # Set up file change monitoring
        detected_changes = []
        
        def on_file_change(events):
            """Callback function to handle file changes."""
            if events:  # Only process if there are actual events
                print(f"\n🔔 Detected {len(events)} file changes:")
                for event in events:
                    print(f"   📄 {event.event_type.upper()}: {event.path} ({event.path_type})")
                detected_changes.extend(events)
        
        print(f"\n👁️  Starting directory monitoring...")
        print("   Press Ctrl+C to stop monitoring")
        
        # Start monitoring
        monitor_thread = session.file_system.watch_directory(
            path=test_dir,
            callback=on_file_change,
            interval=1.0  # Check every second
        )
        monitor_thread.start()
        print("✅ Directory monitoring started")
        
        # Demonstrate file operations
        print("\n🔨 Demonstrating file operations...")
        
        # Create some files
        files_to_create = [
            ("example1.txt", "Hello, World!"),
            ("example2.txt", "This is a test file."),
            ("config.json", '{"setting": "value"}')
        ]
        
        for filename, content in files_to_create:
            filepath = f"{test_dir}/{filename}"
            print(f"   Creating: {filename}")
            
            write_result = session.file_system.write_file(filepath, content)
            if write_result.success:
                print(f"   ✅ Created: {filename}")
            else:
                print(f"   ❌ Failed to create {filename}: {write_result.error_message}")
            
            time.sleep(1.5)  # Give time for monitoring to detect changes
        
        # Modify a file
        print("\n   Modifying example1.txt...")
        modify_result = session.file_system.write_file(
            f"{test_dir}/example1.txt", 
            "Hello, World! - Modified content"
        )
        if modify_result.success:
            print("   ✅ Modified example1.txt")
        else:
            print(f"   ❌ Failed to modify file: {modify_result.error_message}")
        
        # Wait a bit more to capture all events
        print("\n⏳ Waiting for final events...")
        time.sleep(3)
        
        # Stop monitoring
        print("\n🛑 Stopping directory monitoring...")
        monitor_thread.stop_event.set()
        monitor_thread.join(timeout=5)
        print("✅ Directory monitoring stopped")
        
        # Summary
        print(f"\n📊 Summary:")
        print(f"   Total events detected: {len(detected_changes)}")
        
        if detected_changes:
            print("   Event breakdown:")
            event_types = {}
            for event in detected_changes:
                event_types[event.event_type] = event_types.get(event.event_type, 0) + 1
            
            for event_type, count in event_types.items():
                print(f"     {event_type}: {count}")
        
        print("\n✨ Example completed successfully!")
        
    except KeyboardInterrupt:
        print("\n\n⏹️  Monitoring stopped by user")
        
    except Exception as e:
        print(f"\n❌ An error occurred: {e}")
        
    finally:
        # Clean up session
        print("\n🧹 Cleaning up...")
        delete_result = agentbay.delete(session)
        if delete_result.success:
            print("✅ Session cleaned up successfully")
        else:
            print(f"⚠️  Session cleanup warning: {delete_result.error_message}")


if __name__ == "__main__":
    main() 

File: typescript/docs/examples/common-features/basics/filesystem-example/watch-directory-example.ts
/**
 * Watch Directory Example
 * 
 * This example demonstrates how to use the watch_directory functionality
 * to monitor file changes in a directory.
 */

import { AgentBay, CreateSessionParams } from "wuying-agentbay-sdk";
import { FileChangeEvent } from "wuying-agentbay-sdk";

async function main(): Promise<void> {
  // Get API key from environment variable
  const apiKey = process.env.AGENTBAY_API_KEY;
  if (!apiKey) {
    console.log("❌ Please set the AGENTBAY_API_KEY environment variable");
    return;
  }

  console.log("🚀 Watch Directory Example");
  console.log("=".repeat(50));

  try {
    // Initialize AgentBay client
    const agentbay = new AgentBay({ apiKey });
    console.log("✅ AgentBay client initialized");

    // Create session with code_latest ImageId
    const sessionParams = { imageId: "code_latest" };
    const sessionResult = await agentbay.create(sessionParams);

    if (!sessionResult.success) {
      console.log(`❌ Failed to create session: ${sessionResult.errorMessage}`);
      return;
    }

    const session = sessionResult.session;
    console.log(`✅ Session created: ${session.getSessionId()}`);

    try {
      // Create a test directory to monitor
      const testDir = "/tmp/watch_example";
      console.log(`\n📁 Creating test directory: ${testDir}`);

      const createResult = await session.fileSystem.createDirectory(testDir);
      if (!createResult.success) {
        console.log(`❌ Failed to create directory: ${createResult.errorMessage}`);
        return;
      }

      console.log("✅ Test directory created");

      // Set up file change monitoring
      const detectedChanges: FileChangeEvent[] = [];

      const onFileChange = (events: FileChangeEvent[]) => {
        console.log(`\n🔔 Detected ${events.length} file changes:`);
        for (const event of events) {
          console.log(`   📄 ${event.eventType.toUpperCase()}: ${event.path} (${event.pathType})`);
        }
        detectedChanges.push(...events);
      };

      console.log("\n👁️  Starting directory monitoring...");
      console.log("   Press Ctrl+C to stop monitoring");

      // Start monitoring
      const controller = new AbortController();
      const { monitoring, ready } = session.fileSystem.watchDirectory(
        testDir,
        onFileChange,
        1000, // Check every second
        controller.signal
      );

      console.log("✅ Directory monitoring started");

      // Wait for baseline to be established before performing file operations
      await ready;

      // Demonstrate file operations
      console.log("\n🔨 Demonstrating file operations...");

      // Create some files
      const filesToCreate = [
        { name: "example1.txt", content: "Hello, World!" },
        { name: "example2.txt", content: "This is a test file." },
        { name: "config.json", content: '{"setting": "value"}' }
      ];

      for (const file of filesToCreate) {
        const filepath = `${testDir}/${file.name}`;
        console.log(`   Creating: ${file.name}`);

        const writeResult = await session.fileSystem.writeFile(filepath, file.content);
        if (writeResult.success) {
          console.log(`   ✅ Created: ${file.name}`);
        } else {
          console.log(`   ❌ Failed to create ${file.name}: ${writeResult.errorMessage}`);
        }

        // Give time for monitoring to detect changes
        await new Promise(resolve => setTimeout(resolve, 1500));
      }

      // Modify a file
      console.log("\n   Modifying example1.txt...");
      const modifyResult = await session.fileSystem.writeFile(
        `${testDir}/example1.txt`,
        "Hello, World! - Modified content"
      );
      if (modifyResult.success) {
        console.log("   ✅ Modified example1.txt");
      } else {
        console.log(`   ❌ Failed to modify file: ${modifyResult.errorMessage}`);
      }

      // Wait a bit more to capture all events
      console.log("\n⏳ Waiting for final events...");
      await new Promise(resolve => setTimeout(resolve, 3000));

      // Stop monitoring
      console.log("\n🛑 Stopping directory monitoring...");
      controller.abort();
      await monitoring;
      console.log("✅ Directory monitoring stopped");

      // Summary
      console.log("\n📊 Summary:");
      console.log(`   Total events detected: ${detectedChanges.length}`);

      if (detectedChanges.length > 0) {
        console.log("   Event breakdown:");
        const eventTypes: Record<string, number> = {};
        for (const event of detectedChanges) {
          eventTypes[event.eventType] = (eventTypes[event.eventType] || 0) + 1;
        }

        for (const [eventType, count] of Object.entries(eventTypes)) {
          console.log(`     ${eventType}: ${count}`);
        }
      }

      console.log("\n✨ Example completed successfully!");

    } finally {
      // Clean up session
      console.log("\n🧹 Cleaning up...");
      const deleteResult = await agentbay.delete(session);
      if (deleteResult.success) {
        console.log("✅ Session cleaned up successfully");
      } else {
        console.log(`⚠️  Session cleanup warning: ${deleteResult.errorMessage}`);
      }
    }

  } catch (error) {
    if (error instanceof Error && error.name === 'AbortError') {
      console.log("\n\n⏹️  Monitoring stopped by user");
    } else {
      console.log(`\n❌ An error occurred: ${error}`);
    }
  }
}

// Handle Ctrl+C gracefully
process.on('SIGINT', () => {
  console.log("\n\n⏹️  Received interrupt signal, stopping...");
  process.exit(0);
});

main().catch(console.error); 